microcloud/cmd/microcloud/ask.go

```diff
@@ -69,6 +70,75 @@ func askMissingServices(services []types.ServiceType, stateDirs map[types.Servic
 	return services, nil
 }
 
+func askAddress(autoSetup bool, listenAddr string) (string, *net.IPNet, error) {
+	if listenAddr == "" {
+		info, err := mdns.GetNetworkInfo()
+		if err != nil {
+			return "", nil, fmt.Errorf("Failed to find network interfaces: %w", err)
+		}
+
+		if len(info) == 0 {
+			return "", nil, fmt.Errorf("Found no valid network interfaces")
+		}
+
+		listenAddr = info[0].Address
+		if !autoSetup && len(info) > 1 {
+			data := make([][]string, 0, len(info))
+			for _, net := range info {
+				data = append(data, []string{net.Address, net.Interface})
+			}
+
+			table := NewSelectableTable([]string{"ADDRESS", "IFACE"}, data)
+			askRetry("Retry selecting an address?", autoSetup, func() error {
+				fmt.Println("Select an address for MicroCloud's internal traffic")
+				table.Render(table.rows)
+				answers, err := table.GetSelections()
+				if err != nil {
+					return err
+				}
+
+				if len(answers) != 1 {
+					return fmt.Errorf("Please select exactly one address")
+				}
+
+				listenAddr = table.SelectionValue(answers[0], "ADDRESS")
+
+				fmt.Printf(" Using address %q for MicroCloud\n\n", listenAddr)
+
+				return nil
+			})
+		} else {
+			fmt.Printf("Using address %q for MicroCloud\n", listenAddr)
+		}
+
+	}
+
+	var subnet *net.IPNet
+	var err error
+	if net.ParseIP(listenAddr).To4() != nil {
+		_, subnet, err = net.ParseCIDR(listenAddr + "/24")
+	} else {
+		_, subnet, err = net.ParseCIDR(listenAddr + "/64")
+	}
```

Comment:
Can we get the CIDR from the address instead of assuming /24 for v4 and /64 for v6 as that can be very wrong?

I'd probably have gone with `net.InterfaceAddrs()` which gets you CIDR of all addresses on the system.
You'd then need to filter our anything that's not a global address from that by using ip.ParseIP and only going for GlobalUnicast.

Reply:
Yeah that's a good point. I was less concerned about invalid subnets since I had added the  `--subnet` flag, but since that's gone it can be problematic.

I've added a `Subnet` field to the `NetworkInfo` struct now. So any selected address will have its associated subnet be used. If the user specifies their own address with `--address`, init will fail if that address doesn't fit in any of the found subnets.